### PR TITLE
[WIP] write references to mutable pointers in data and tls segment

### PR DESCRIPTION
**Pull request by rainers**
While preparing a new release of Visual D, I noticed that the (rebuilt) dmd version I used so far (2.066 with the precise GC from DConf 2013) generates corrupt DLLs. Instead of trying to figure the reason for this, I tried the precise GC from https://github.com/dlang/druntime/pull/1603, but Visual D still suffers a lot when editing std.datetime with it: there are too many false pointers in the data segment.

With this PR I reimplemented a simple version of the additional info required to scan the data and tls segments precisely: instead of storing type information, it just lists addresses of all mutable pointers. This even makes usage of the heap-precise GC optional.

For example, the phobos unittest executable has a data segment of about 3 MB (and 11 kB TLS), but only 262 (and 282) mutable pointers. So this PR currently adds about 2kB to the data segment.

This PR only implements emission for Windows OMF and COFF, but should be portable to other platforms pretty easily.

### CODE REVIEW

**Comment by rainers** [comment]
Corresponding druntime PR: https://github.com/dlang/druntime/pull/1763

File: src/backend/dt.c [hunk 1]
```diff
@@ -411,8 +411,9 @@ dt_t *DtBuilder::xoffpatch(Symbol *s, unsigned offset, tym_t ty)
 
 /*************************************
  * Create a reference to another dt.
+ * Return the internal symbol used for the other dt
```
Comment (WalterBright) [inline src/backend/dt.c]: Use proper Ddoc formatting, i.e. `Returns:`

File: src/toobj.d [hunk 2]
```diff
@@ -220,6 +221,41 @@ void genModuleInfo(Module m)
     objmod.moduleinfo(msym);
 }
 
+void write_pointers(Type type, Symbol *s, uint offset)
```
Comment (WalterBright) [inline src/toobj.d]: No documentation for function.

File: src/traits.d [hunk 3]
```diff
@@ -112,48 +112,29 @@ static this()
 /**
  * get an array of size_t values that indicate possible pointer words in memory
  *  if interpreted as the type given as argument
- * the first array element is the size of the type for independent interpretation
- *  of the array
- * following elements bits represent one word (4/8 bytes depending on the target
- *  architecture). If set the corresponding memory might contain a pointer/reference.
- *
- *  [T.sizeof, pointerbit0-31/63, pointerbit32/64-63/128, ...]
+ * returns the size of the type in bytes, d_uns64.max on error
```
Comment (WalterBright) [inline src/traits.d]: Use proper Ddoc formatting, `Returns:`

File: src/traits.d [hunk 4]
```diff
@@ -112,48 +112,29 @@ static this()
...
  */
-extern (C++) Expression pointerBitmap(TraitsExp e)
+extern (C++) d_uns64 getTypePointerBitmap(Loc loc, Type t, Array!(d_uns64)* data)
```
Comment (WalterBright) [inline src/traits.d]: Document arguments

File: src/traits.d [hunk 5]
```diff
@@ -112,48 +112,29 @@ static this()
...
     d_uns64 sz;
     if (t.ty == Tclass && !(cast(TypeClass)t).sym.isInterfaceDeclaration())
-        sz = (cast(TypeClass)t).sym.AggregateDeclaration.size(e.loc);
+        sz = (cast(TypeClass)t).sym.AggregateDeclaration.size(loc);
     else
-        sz = t.size(e.loc);
+        sz = t.size(loc);
     if (sz == SIZE_INVALID)
-        return new ErrorExp();
+        return d_uns64.max;
```
Comment (WalterBright) [inline src/traits.d]: This seems outside the scope of this PR.

File: src/traits.d [hunk 6]
```diff
@@ -338,21 +319,51 @@ extern (C++) Expression pointerBitmap(TraitsExp e)
...
+/**
+ * get an array of size_t values that indicate possible pointer words in memory
+ *  if interpreted as the type given as argument
+ * the first array element is the size of the type for independent interpretation
+ *  of the array
+ * following elements bits represent one word (4/8 bytes depending on the target
+ *  architecture). If set the corresponding memory might contain a pointer/reference.
+ *
+ *  [T.sizeof, pointerbit0-31/63, pointerbit32/64-63/128, ...]
+ */
```
Comment (WalterBright) [inline src/traits.d]: Missing documentation on the return value.

**Comment by WalterBright** [comment]
General comment: many added functions have zero documentation. Please follow Ddoc conventions for this, even in the `.c` files, since they'll be converted to D sooner or later.

**Comment by WalterBright** [comment]
What is the effect of this on compilation speed?

**Comment by rainers** [comment]
Updated comments. I've been following the (missing) documentation style of functions nearby ;-)

File: src/traits.d (hunk 5 above)
Comment (rainers) [inline src/traits.d]: There is no functional change here. getTypePointerBitmap() has just been extracted from pointerBitmap() to be callable from elsewhere. Having to deal with the Expression return value in write_instance_pointers() would be wasteful.

**Comment by rainers** [comment]
> What is the effect of this on compilation speed?

I haven't measured it yet, but I don't expect much. The phobos unittest build emits less than 600 pointer references with a data segment of about 3 MB.

I was conidering caching the result of getTypePointerBitmap() and hasPointers() in the type, though.

File: src/backend/cgobj.c [hunk 7]
```diff
@@ -3702,5 +3708,94 @@ void Obj::gotref(Symbol *s)
...
+ */
+
+void Obj::write_pointerRef(Symbol* s, unsigned soff)
+{
+    if (!obj.ptrref_buf)
+        obj.ptrref_buf = new Outbuffer;
+
+    // defer writing pointer references until the symbols are written out
+    obj.ptrref_buf->write(&s, sizeof(s));
+    obj.ptrref_buf->write(&soff, sizeof(soff));
```
Comment (WalterBright) [inline src/backend/cgobj.c]: Can use `write32()` here.

File: src/backend/cgobj.c [hunk 8]
```diff
@@ -3702,5 +3708,94 @@ void Obj::gotref(Symbol *s)
 {
 }
 
+/*****************************************
+ * write a reference to a mutable pointer into the object file
+ * Input:
```
Comment (WalterBright) [inline src/backend/cgobj.c]: The Ddoc way is:
```
Params:
    s = symbol that contains the pointer
    soff = offset of the pointer inside the symbol's memory
```

Comment (WalterBright) [inline src/backend/cgobj.c]: This applies throughout the PR. Also, `symbols'` should be `Symbol's` globally to this PR.

**Comment by WalterBright** [comment]
Thanks, your added comments have improved my comprehension of this greatly. I noticed that you put the tables in readonly sections in COFF. This can be done for OMF too if they are put in code sections. I'm currently doing that with the string literal output I've been re-implementing, it seems to work well. The advantage of RO segments is that multiple users of the DLL won't need extra copies of it, and the GC should not be scanning them.

Also, for data emitted to RO sections, there should not be pointer tables generated.

**Comment by WalterBright** [comment]
> I've been following the (missing) documentation style of functions nearby ;-)

Unfortunately, those functions predate Ddoc and the style is not recognized by Ddoc. They will eventually get fixed, but new stuff should follow Ddoc style.

**Comment by rainers** [comment]
Thanks for review, I've updated the comments.

> The advantage of RO segments is that multiple users of the DLL won't need extra copies of it

You mean multiple processes using the same physical pages? Due to relocations inside the text segment and ALSR, it's unlikely to never be modified.

> and the GC should not be scanning them.

With the generated information used, they won't be scanned anyway. But indeed, it wouldn't hurt to exclude them for the generic case, too.

> Also, for data emitted to RO sections, there should not be pointer tables generated.

The hope is that this data should be typed as immutable/const, so it fails the `isMutable` tests. Is this assumption wrong?

**Comment by rainers** [comment]
> But indeed, it wouldn't hurt to exclude them for the generic case, too.

Now moved the data into the CODE segment.

**Comment by WalterBright** [comment]
For readonly sections, yes, multiple process can use the same physical pages.

> The hope is that this data should be typed as immutable/const, so it fails the isMutable tests. Is this assumption wrong?

Some of the RO data is not emitted via typed data. I plan to look into emitting all immutable statically inited data into RO sections. The GC should just skip those sections, as they can never contain pointers to mutable data.

Thanks for making the other fixes.

**Comment by rainers** [comment]
Hmm, thanks for merging, but I didn't expect that ATM. win32.mak was changed to pull in a druntime PR for better testing...

**Comment by rainers** [comment]
See https://github.com/dlang/dmd/pull/6549. The appropriate druntime PR is https://github.com/dlang/druntime/pull/1763

**Comment by MartinNowak** [comment]
This is a different approach than [Issue 14472 – add separate ptr data section](https://issues.dlang.org/show_bug.cgi?id=14472), can we discuss the pros/cons of those? I'd assume that the memory access pattern of those 262 pointers in 3MiB is a mess and not prefetchable, while simply moving all data with mutable pointers (e.g. anything that needs GC scanning) shouldn't be much bigger.
Also the double dereference has an additional overhead.

**Comment by rainers** [comment]
> This is a different approach than Issue 14472 – add separate ptr data section, can we discuss the pros/cons of those?

Adding immutable data or data without pointers into a separate sections works too, but doesn't work for struct with a mixture of pointers and non-pointers. The version from DConf2013 linked in the bugzilla report saves full type information for variables with pointers (but ignored indirections inside mutable initializers, e.g. multi-dimensional arrays or class instances in the data segment). The approach in this PR is simpler, as it doesn't create more dependencies to type information. My impression is that it even needs less additional memory. I don't have a working version of the other implementation to compare, though.

> I'd assume that the memory access pattern of those 262 pointers in 3MiB is a mess and not prefetchable, while simply moving all data with mutable pointers (e.g. anything that needs GC scanning) shouldn't be much bigger.

The current GC sorts ranges, so I expect the accesses not to be worse than scanning the full 3MB. One possible example of a bad data structure is a static array of strings, where you have a series of length and pointer pairs.

> Also the double dereference has an additional overhead.

You mean the GC range pointing to the pointer? I think that's pretty unavoidable.

**Comment by MartinNowak** [comment]
It's a good start in any case.

> but doesn't work for struct with a mixture of pointers and non-pointers

Sure, but that could be enhanced by precise scan information.

Can we at least get the number, what size would it be if we moved all data _containing_ a mutable pointer into a separate section.

> You mean the GC range pointing to the pointer? I think that's pretty unavoidable.

I mean an array of pointers each pointing to the `.ptr` of an array. In particular the "static array of strings" is an example where conservatively (or precisely) scanning that array should be faster than iterating a separate array pointing at every 2nd element.

**Comment by rainers** [comment]
> Can we at least get the number, what size would it be if we moved all data containing a mutable pointer into a separate section.

I've added some output to dmd to print some statistics, this yields (modulo duplicates due to COMDATs in separate object files) for the phobos unittest:

* 70 data items with pointers only
* 357 data items with a mixture of pointers and non-pointers
* these contain 456 pointers in 995 words.

Looks very much like a lot of slices in global/tls memory.
